### ISSUE TITLE
Direct Future Prediction

### ISSUE THREAD

**Issue by dmadeka** [NervanaSystems/coach#227]
I get a strange error when I try to run DFP on a custom environment (with a Discrete action space).

```python
AttributeError                            Traceback (most recent call last)
<ipython-input-35-2055451d4b4a> in <module>()
     22     agent_params=agent_params,
     23     env_params=env_params,
---> 24     schedule_params=SimpleSchedule()
     25 )

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/rl_coach/graph_managers/basic_rl_graph_manager.py in __init__(self, agent_params, env_params, schedule_params, vis_params, preset_validation_params)
     39 
     40         self.agent_params.visualization = vis_params
---> 41         if self.agent_params.input_filter is None:
     42             self.agent_params.input_filter = env_params.default_input_filter()
     43         if self.agent_params.output_filter is None:

AttributeError: 'DFPAlgorithmParameters' object has no attribute 'input_filter'
```

The invocation is as follows:

```python
# define the environment parameters
bit_length = 10
env_params = GymVectorEnvironment(level='./custom.py')
env_params.additional_simulator_parameters = { 'num_states': 100}

agent_params = DFPAlgorithmParameters()

graph_manager = BasicRLGraphManager(
    agent_params=agent_params,
    env_params=env_params,
    schedule_params=SimpleSchedule()
)
```

**Comment by dmadeka** [comment]
Bumping this up

**Comment by dmadeka** [comment]
@scttl If there's a way for me to contribute back here - happy to do so, not sure where to start though

**Comment by galleibo-intel** [comment]
There's an issue with the preset code, as the `agent_params` variable in the code is pointing at an instance of `AlgorithmParameters` instead of an `AgentParameters` instance. Note that the instantiation of `BasicRLGraphManager` expects an instance of `AgentParameters`.

**Comment by dmadeka** [comment]
@galleibo-intel If i pass an `DFPAgentParameters` instance - it raises the error

```python
~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/rl_coach/agents/dfp_agent.py in __init__(self, agent_parameters, parent)
    138 class DFPAgent(Agent):
    139     def __init__(self, agent_parameters, parent: Union['LevelManager', 'CompositeAgent']=None):
--> 140         super().__init__(agent_parameters, parent)
    141         self.current_goal = self.ap.algorithm.goal_vector
    142         self.target_measurements_scale_factors = None

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/rl_coach/agents/agent.py in __init__(self, agent_parameters, parent)
    112         # filters
    113         self.input_filter = self.ap.input_filter
--> 114         self.input_filter.set_name('input_filter')
    115         self.output_filter = self.ap.output_filter
    116         self.output_filter.set_name('output_filter')

AttributeError: 'NoneType' object has no attribute 'set_name'
```

**Comment by galleibo-intel** [comment]
My guess is that your custom environment does not define a `default_input_filter` (take a look at any of the built-in environments, e.g. Doom). If there's no filtering needed, just set an empty input filter. So either add default input and output filters, by adding to your environment's init code - 
```
        self.default_input_filter = NoInputFilter()
        self.default_output_filter = NoOutputFilter()
```
Or just define one from your preset.  by adding to your preset - 

```
       agent_params.input_filter = NoInputFilter()
       agent_params.output_filter = NoOutputFilter()
```

**Comment by dmadeka** [comment]
My environment is just an OpenAI gym environment. Not sure that those have `default_input_filter` as a necessity?

If I try to set those directly - I get the following error:

```python
~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/training/checkpointable/base.py in _add_variable_with_custom_getter(self, name, shape, dtype, initializer, getter, overwrite, **kwargs_for_getter)
    637     new_variable = getter(
    638         name=name, shape=shape, dtype=dtype, initializer=initializer,
--> 639         **kwargs_for_getter)
    640 
    641     # If we set an initializer and the variable processed it, tracking will not

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/ops/variable_scope.py in get_variable(name, shape, dtype, initializer, regularizer, trainable, collections, caching_device, partitioner, validate_shape, use_resource, custom_getter, constraint, synchronization, aggregation)
   1485       constraint=constraint,
   1486       synchronization=synchronization,
-> 1487       aggregation=aggregation)
   1488 
   1489 

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/ops/variable_scope.py in get_variable(self, var_store, name, shape, dtype, initializer, regularizer, reuse, trainable, collections, caching_device, partitioner, validate_shape, use_resource, custom_getter, constraint, synchronization, aggregation)
   1235           constraint=constraint,
   1236           synchronization=synchronization,
-> 1237           aggregation=aggregation)
   1238 
   1239   def _get_partitioned_variable(self,

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/ops/variable_scope.py in get_variable(self, name, shape, dtype, initializer, regularizer, reuse, trainable, collections, caching_device, partitioner, validate_shape, use_resource, custom_getter, constraint, synchronization, aggregation)
    538           constraint=constraint,
    539           synchronization=synchronization,
--> 540           aggregation=aggregation)
    541 
    542   def _get_partitioned_variable(self,

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/ops/variable_scope.py in _true_getter(name, shape, dtype, initializer, regularizer, reuse, trainable, collections, caching_device, partitioner, validate_shape, use_resource, constraint, synchronization, aggregation)
    490           constraint=constraint,
    491           synchronization=synchronization,
--> 492           aggregation=aggregation)
    493 
    494     # Set trainable value based on synchronization value.

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/ops/variable_scope.py in _get_single_variable(self, name, shape, dtype, initializer, regularizer, partition_info, reuse, trainable, collections, caching_device, validate_shape, use_resource, constraint, synchronization, aggregation)
    859                          "reuse=tf.AUTO_REUSE in VarScope? "
    860                          "Originally defined at:\n\n%s" % (
--> 861                              name, "".join(traceback.format_list(tb))))
    862       found_var = self._vars[name]
    863       if not shape.is_compatible_with(found_var.get_shape()):

ValueError: Variable main_level/agent/main/online/network_0/goal/Dense_0/kernel already exists, disallowed. Did you mean to set reuse=True or reuse=tf.AUTO_REUSE in VarScope? Originally defined at:

  File "/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/rl_coach/architectures/tensorflow_components/layers.py", line 144, in __call__
    activation=activation)
  File "/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/rl_coach/architectures/tensorflow_components/embedders/embedder.py", line 115, in _build_module
    is_training=self.is_training)
  File "/home/ubuntu/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/rl_coach/architectures/tensorflow_components/embedders/embedder.py", line 88, in __call__
    self._build_module()
```

**Comment by dmadeka** [comment]
Hmm, looks like the entire agent is only built for images. Not entirely sure why -- it should generalize to the feed forward case. Is there a simple example to override this kind of config?

**Comment by galleibo-intel** [comment]
An image embedder is created in case the environment's state space is an image. If it's a vector, a vector embedder is created.

Anyway, this discussion is a bit out of scope for this issue.
 If another issue (in Coach) arises, feel free to file a new issue. Obviously, an issue should be filed only after you have made sure that this is an issue in Coach, and not in the user's code. Reference code reproducing the issue would also help. There are many example presets, and quite a few example environments for you to look at.
